adds latest requirements

junehoe opened the pull request:
Addresses current PR feedback, adds ability to choose computer player difficulty, adds unbeatable computer, and implements the custom Minimax Ruby gem.

klkelley commented on README.md:
````diff
@@ -0,0 +1,44 @@
+Inside of the  directory, run:
+```
+ruby lib/play.rb
+```
````
I had errors following these directions, I had to run `bundle install` before being able to run the program

klkelley replied:
In addition to direction on how to play the game, you might add something on how to run the tests as well.

klkelley commented on README.md:
````diff
@@ -0,0 +1,44 @@
+```
+
+## How to Contribute
+1. Fork this repository to your account.
````
I like that you added this contribution section!

klkelley commented on lib/console_io.rb:
```diff
@@ -1,18 +1,52 @@
+  end
+
+  def display_greeting_message
+    display_message("Welcome to Ruby tic-tac-toe!")
```
I like that all of the messages are contained in one place but what if you had to support multiple languages? (Doesn't need to be implemented, just something to think about)

klkelley commented on lib/game.rb:
```diff
@@ -1,71 +1,52 @@
-    board.get_columns.each { |column| return true if winning_combo?(column) }
-    false
+  def game_over?
+    game_logic.winner? or game_logic.draw?
```
It seems the `GameLogic` class also has this same logic and I think it makes more sense for it to be the responsibility of the `GameLogic` class as well.

klkelley commented on lib/play.rb:
```diff
@@ -1,10 +1,32 @@
+  easy_player_2: easy_player_2,
+  unbeatable_player_2: unbeatable_player_2
+})
+player_initializer.setup_players(3)
```
What is `3` here?

klkelley commented:
I left some comments, it was a bit difficult to know what code was part of PR feedback and the new feature but if you have any questions let me know.

junehoe commented:
Thank you for the feedback Kara! I have addressed your feedback immediately:

-Added `bundle install` and `bin/rspec` commands to README
-`game_over` method removed from `Game` and now uses `game_over` from `GameLogic`
-Removes the magic number 3 and sets it to a constant within `PlayerInitializer`. 3 was supposed to be the number of selections to initialize the players to run the game

omarnyte commented on lib/game.rb:
```diff
@@ -1,71 +1,48 @@
-    if winner?
-      "#{curr_player.name} wins"
+    console_io.display_greeting_message
+    until game_logic.game_over?
```
I really like that the code for determining when the game is over has been moved to the GameLogic class. It really helped clean up your game_spec because you no longer need to write out any winning/draw combinations in any of its tests. Instead, you can test the behavior of that class given that the game is over or not without needing to know how the game is determined to be over.

omarnyte commented on lib/game.rb:
```diff
@@ -1,71 +1,48 @@
-    @console_io = console_io
+  def initialize args
+    @board = args[:board]
+    @console_io = args[:console_io]
```
I think you've done a pretty good job injecting dependences into your Game. In fact, I'd say that it'd be relatively easy to have swap in some other *type* of input/output (so long as that new input/output [walks like a ConsoleIO and quacks like a ConsoleIO](https://en.wikipedia.org/wiki/Duck_typing)). That being said, do you think that the instance variable name `console_io` would age well in your system?

omarnyte commented:
I really like the direction in which you've taken your game. I second Kara's comments, but I don't have any major concerns. Good job!

junehoe commented:
The feedback received up until now has been implemented.

jayden commented:
I agree with Kara that it is difficult to tell what changes address feedback vs implements the new feature. Moving forward, let's address feedback in the currently opened PR to avoid this kind of confusion.

Something that I also want you to start thinking about is your commit messages and how they are worded/organized. For example, commit messages like `implements omar's feedback` isn't very meaningful to add to your git history because it doesn't specifically tell me what has changed in the project. [This blogpost](https://chris.beams.io/posts/git-commit/) is pretty good at laying out some guidelines.

jayden commented on lib/play.rb:
```diff
@@ -1,10 +1,35 @@
+players = {
+  human_player_1: HumanPlayer,
+  human_player_2: HumanPlayer,
+  easy_player_2: EasyPlayer,
```
If I were to ask you to implement another computer player of medium difficulty, how many places in your codebase would need to change to add this new requirement?

If we take a look at both the Easy and Unbeatable player classes, the only difference between the two is the AI logic itself. What if instead of creating different computer player classes that depend on AI logic, we `invert` the dependencies and inject them into the class? [Here's a great article](https://medium.com/@joshsaintjacque/ruby-the-strategy-pattern-16c98b99b373) describes the Strategy pattern in Ruby.

jayden commented on lib/play.rb:
```diff
@@ -1,10 +1,35 @@
+  board: ttt_board,
+  ui: console_io,
+  game_logic: game_logic,
+  curr_player: player_initializer.ttt_players[PLAYER_1],
```
It seems odd that game needs to be told who the `curr_player` and `next_player` are. What do you think about just passing the array of players to the game and let it determine whose turn it is?
